kernel/src/kernel.rs

```diff
@@ -41,10 +46,43 @@ use crate::utilities::cells::NumericCellExt;
 /// is less than this threshold.
 pub(crate) const MIN_QUANTA_THRESHOLD_US: u32 = 500;
 
+#[derive(Copy, Clone)]
+pub struct StaticSlice<T> {
```

Comment:
?

Reply:
A `[u8]` is not `Sized`, so can't be stored in a bunch of places. In Rust `std` statics (like `thread_local`) end up using `Box` or `Vec`. `StaticSlice` is basically the a fixed-sized `Vec` (it has a fixed length, but that length is stored dynamically), so it is `Sized` and can be stored in statics directly, like a fixed-sized array.

Reply:
I'm having trouble seeing the semantic difference between a `StaticSlice<T>` and a `&'static mut [T]`.

Reply:
There should be no semantic difference, only that `&'static mut [T]` is not `Sized` so cannot be declared in a context that needs `Sized` (such as a a static variable).

Reply:
Ah, I see. The issue is not `Sized` but that `mut` references are not allowed in a `static` item's type. Maybe a different name and/or a link to https://github.com/rust-lang/rust/issues/57349 would make this more clear?

Some other improvement ideas:

1. `StaticSlice` doesn't actually need to be slice-specific. You can make it generic for any type and let the user choose whether to store a slice in it.
2. You can store a `&'static mut T` instead of a raw pointer to avoid `unsafe` in `Deref` and `DerefMut`.

Demonstration: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8de22b70edcbedaffccc382fd35ddd11